nixos/redis: add support for more auth options

ashleyghooper opened the pull request:
The current Redis module is missing some options that are important for running a secure Redis infrastructure, including:

- `masteruser` and `masterauth` for authentication between replicas and master
- `sentinel auth-user` and `sentinel auth-pass` to provide credentials that a Sentinel instance should use to authenticate to the Redis instance it is monitoring

This PR adds the above options, along with the Sentinel master name, port, and quorum number that are required for sentinel to monitor a Redis instance. New assertions have also been added to prevent configuration issues before build/deployment.

A working example Redis + Sentinel configuration using this updated module is below:

```nix
{
  config,
  hosts,
  hostSpec,
  lib,
  pkgs,
  ...
}: let
  isReplica = hostSpec.hostname != valkeyPrimary.hostname;
  redisInstance = "valkey-cluster";
  redisPort = 6379;
  sentinelPort = 26379;
  valkeyPrimary = hosts.${hostSpec.config.valkey.primary};
in {
  sops = {
    secrets = {
      "valkey/servers/main/aclfile_content" = {
        owner = "redis-main";
        group = "redis-main";
        restartUnits = ["redis-main.service"];
      };
      "valkey/servers/main/users/master/password" = {
        restartUnits = ["redis-main.service" "redis-sentinel.service"];
      };
      "valkey/servers/sentinel/users/sentinel/password" = {
        group = "redis-sentinel";
        owner = "redis-sentinel";
        restartUnits = ["redis-sentinel.service"];
      };
    };
  };

  services.redis = {
    package = pkgs.valkey;
    servers = {
      "main" = {
        bind = "0.0.0.0";
        enable = true;
        openFirewall = true;
        port = redisPort;
        masterUser = "master";
        masterAuthFile = config.sops.secrets."valkey/servers/main/users/master/password".path;
        requirePassFile = config.sops.secrets."valkey/servers/main/users/master/password".path;
        settings =
          {
            # Security
            aclfile = config.sops.secrets."valkey/servers/main/aclfile_content".path;
            protected-mode = "no";
            # Persistence
            appendonly = "yes";
          }
          // (lib.optionalAttrs isReplica {
            replicaof = "${valkeyPrimary.ipv4Primary.address} ${toString redisPort}";
          });
      };
      "sentinel" = {
        bind = "0.0.0.0";
        enable = true;
        openFirewall = true;
        port = sentinelPort;
        extraParams = ["--sentinel"];
        sentinelAuthUser = "sentinel";
        sentinelAuthPassFile = config.sops.secrets."valkey/servers/sentinel/users/sentinel/password".path;
        sentinelMasterName = redisInstance;
        sentinelMasterPort = redisPort;
        sentinelMasterQuorum = 2;
        settings = {
          protected-mode = "no";
        };
      };
    };
  };
}
```

## Things done

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [x] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

ashleyghooper committed:
nixos/redis: add support for more auth options

- Add Redis auth options that are important for running a
secure Redis infrastructure
- `masteruser` and `masterauth` for authentication between
replicas and master
- `sentinel auth-user` and `sentinel auth-pass` to provide
credentials that a Sentinel instance should use to
authenticate to the Redis instance it is monitoring
- Add assertions to catch configuration issues before
build/deployment

Signed-off-by: Ashley Hooper <ashleyghooper@gmail.com>

SuperSandro2000 commented on nixos/modules/services/databases/redis.nix:
```diff
@@ -338,6 +338,21 @@ in
+                    default = null;
+                    description = ''
+                      If the master is password protected via ACLs this option can be used to specify
+                      the Redis user that is used by replicas.'';
```
```suggestion
                      the Redis user that is used by replicas.
                    '';
```

SuperSandro2000 reviewed (commented):
otherwise lgtm

ashleyghooper committed:
nixos/redis: rework Sentinel options

- Add sentinelMasterHost, sentinelMasterPort, sentinelMasterQuorum
  options
- Add special handling for mutable state in /var/lib/redis-xxx

Redis/Valkey Sentinel persists dynamic cluster state by rewriting its
configuration file at runtime. This behaviour cannot be disabled.

Add conditional preStart logic to avoid overwriting Sentinel-owned
configuration while preserving NixOS-managed defaults. This introduces
a necessary, documented deviation from strict declarative semantics.
